File: src/api/players.ts
import { api } from "./client";
import type { Player } from "@/types";

interface PlayerFilters {
  region?: string;
  position?: string;
  age_min?: number;
  age_max?: number;
}

interface ApiAnthro {
  height: number;
  weight: number;
  arm_span: number;
  leg_length: number;
  torso_length: number;
  sitting_height: number;
  shoulder_width: number;
  shoe_size: number;
  body_fat_pct?: number | null;
}

interface ApiPlayer {
  id: string;
  owner_id: string;
  first_name: string;
  last_name: string;
  middle_name?: string | null;
  birth_date: string;
  position: string;
  shooting_hand: string;
  city: string;
  region: string;
  team?: string | null;
  jersey_number?: number | null;
  rating: number;
  avatar?: string | null;
  anthropometrics?: ApiAnthro | null;
}

function mapApiPlayer(p: ApiPlayer): Player {
  const anthro = p.anthropometrics;
  return {
    id: p.id,
    firstName: p.first_name,
    lastName: p.last_name,
    middleName: p.middle_name ?? undefined,
    birthDate: p.birth_date,
    position: p.position as Player["position"],
    shootingHand: p.shooting_hand as Player["shootingHand"],
    region: p.region,
    city: p.city,
    team: p.team ?? undefined,
    jerseyNumber: p.jersey_number ?? undefined,
    rating: p.rating,
    avatar: p.avatar ?? "",
    parentId: p.owner_id,
    anthropometrics: anthro
      ? {
          height: anthro.height,
          weight: anthro.weight,
          armSpan: anthro.arm_span,
          legLength: anthro.leg_length,
          torsoLength: anthro.torso_length,
          sittingHeight: anthro.sitting_height,
          shoulderWidth: anthro.shoulder_width,
          shoeSize: anthro.shoe_size,
          bodyFatPct: anthro.body_fat_pct ?? undefined,
        }
      : {
          height: 0,
          weight: 0,
          armSpan: 0,
          legLength: 0,
          torsoLength: 0,
          sittingHeight: 0,
          shoulderWidth: 0,
          shoeSize: 0,
        },
  };
}

export async function fetchPlayers(filters?: PlayerFilters): Promise<Player[]> {
  const params = new URLSearchParams();
  if (filters?.region) params.set("region", filters.region);
  if (filters?.position) params.set("position", filters.position);
  if (filters?.age_min != null) params.set("age_min", String(filters.age_min));
  if (filters?.age_max != null) params.set("age_max", String(filters.age_max));
  const { data } = await api.get<ApiPlayer[]>(`/api/v1/players?${params}`);
  return data.map(mapApiPlayer);
}

export async function fetchPlayer(id: string): Promise<Player | null> {
  try {
    const { data } = await api.get<ApiPlayer>(`/api/v1/players/${id}`);
    return mapApiPlayer(data);
  } catch {
    return null;
  }
}

export interface CreatePlayerBody {
  first_name: string;
  last_name: string;
  middle_name?: string;
  birth_date: string;
  position: string;
  shooting_hand: string;
  city: string;
  region: string;
  team?: string;
  jersey_number?: number;
  anthropometrics: {
    height: number;
    weight: number;
    arm_span: number;
    leg_length: number;
    torso_length: number;
    sitting_height: number;
    shoulder_width: number;
    shoe_size: number;
    body_fat_pct?: number;
  };
}

export async function createPlayer(body: CreatePlayerBody): Promise<Player> {
  const { data } = await api.post<ApiPlayer>("/api/v1/players", body);
  return mapApiPlayer(data);
}

export async function fetchPlayerRating(playerId: string): Promise<{
  rating: number;
  regional_rank: number;
  regional_total: number;
}> {
  const { data } = await api.get(`/api/v1/players/${playerId}/rating`);
  return data;
}

interface ApiTestSession {
  id: string;
  player_id: string;
  recorded_at: string;
  sprint_20m_fwd?: number | null;
  sprint_20m_bwd?: number | null;
  sprint_60m?: number | null;
  standing_jump?: number | null;
  long_jump?: number | null;
  agility?: number | null;
  flexibility?: number | null;
  push_ups?: number | null;
  pull_ups?: number | null;
  plank_sec?: number | null;
  balance_test_sec?: number | null;
}

export async function fetchTestSessions(playerId: string): Promise<ApiTestSession[]> {
  const { data } = await api.get<ApiTestSession[]>(`/api/v1/players/${playerId}/tests`);
  return data;
}

File: backend/app/routers/players.py
from fastapi import APIRouter, Depends, File, HTTPException, Query, UploadFile, status
from sqlalchemy.ext.asyncio import AsyncSession

from app.database import get_db
from app.models.user import User
from app.repositories import player_repo
from app.schemas.player import (
    PlayerCreate,
    PlayerPartialUpdate,
    PlayerUpdate,
    PlayerResponse,
    TestSessionCreate,
    TestSessionResponse,
)
from app.services.auth_service import get_current_user
from app.services import player_service, video_service

MAX_PHOTO_SIZE = 10 * 1024 * 1024
ALLOWED_PHOTO_TYPES = frozenset(
    {"image/jpeg", "image/png", "image/webp"}
)

router = APIRouter(prefix="/api/v1/players", tags=["players"])


@router.post("", response_model=PlayerResponse, status_code=201)
async def create_player(
    body: PlayerCreate,
    db: AsyncSession = Depends(get_db),
    user: User = Depends(get_current_user),
):
    """Create a new player profile (parent only)."""
    player = await player_service.create_player(db, body, user)
    return player


@router.get("", response_model=list[PlayerResponse])
async def list_players(
    region: str | None = Query(None),
    position: str | None = Query(None),
    age_min: int | None = Query(None),
    age_max: int | None = Query(None),
    db: AsyncSession = Depends(get_db),
):
    """List players with optional filters."""
    return await player_repo.list_players(
        db, region=region, position=position, age_min=age_min, age_max=age_max
    )


@router.get("/{player_id}", response_model=PlayerResponse)
async def get_player(player_id: str, db: AsyncSession = Depends(get_db)):
    """Get a single player with anthropometrics."""
    return await player_service.get_player_or_404(db, player_id)


@router.put("/{player_id}", response_model=PlayerResponse)
async def update_player(
    player_id: str,
    body: PlayerUpdate,
    db: AsyncSession = Depends(get_db),
    user: User = Depends(get_current_user),
):
    """Update a player (owner only)."""
    player = await player_service.get_player_or_404(db, player_id)
    player_service.assert_owner(player, user)
    return await player_service.update_player(db, player, body)


@router.patch("/{player_id}", response_model=PlayerResponse)
async def patch_player(
    player_id: str,
    body: PlayerPartialUpdate,
    db: AsyncSession = Depends(get_db),
    user: User = Depends(get_current_user),
):
    player = await player_service.get_player_or_404(db, player_id)
    player_service.assert_owner(player, user)
    return await player_service.patch_player_partial(db, player, body)


@router.post("/{player_id}/photo", status_code=201)
async def upload_player_photo(
    player_id: str,
    file: UploadFile = File(...),
    db: AsyncSession = Depends(get_db),
    user: User = Depends(get_current_user),
):
    player = await player_service.get_player_or_404(db, player_id)
    player_service.assert_owner(player, user)
    if file.content_type not in ALLOWED_PHOTO_TYPES:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Допустимые форматы: jpeg, png, webp",
        )
    content = await file.read()
    if len(content) > MAX_PHOTO_SIZE:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Файл превышает 10MB",
        )
    key = video_service.upload_photo_to_s3(content, file.filename or "photo.jpg")
    player.photo_key = key
    await player_repo.update_player(db, player)
    return {"photo_key": key}


@router.get("/{player_id}/photo")
async def get_player_photo_url(
    player_id: str, db: AsyncSession = Depends(get_db)
):
    player = await player_service.get_player_or_404(db, player_id)
    if not player.photo_key:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Фото не загружено",
        )
    url = video_service.get_presigned_url(player.photo_key)
    return {"url": url, "expires_in": 3600}


@router.delete("/{player_id}", status_code=204)
async def delete_player(
    player_id: str,
    db: AsyncSession = Depends(get_db),
    user: User = Depends(get_current_user),
):
    """Delete a player (owner only)."""
    player = await player_service.get_player_or_404(db, player_id)
    player_service.assert_owner(player, user)
    await player_repo.delete_player(db, player)


@router.get("/{player_id}/rating")
async def get_rating(player_id: str, db: AsyncSession = Depends(get_db)):
    """Get player rating with regional rank."""
    player = await player_service.get_player_or_404(db, player_id)
    rank, total = await player_repo.get_regional_rank(db, player)
    return {
        "player_id": player.id,
        "rating": player.rating,
        "regional_rank": rank,
        "regional_total": total,
    }


@router.post(
    "/{player_id}/tests", response_model=TestSessionResponse, status_code=201
)
async def add_test_session(
    player_id: str,
    body: TestSessionCreate,
    db: AsyncSession = Depends(get_db),
    user: User = Depends(get_current_user),
):
    """Add a physical test session (owner only)."""
    player = await player_service.get_player_or_404(db, player_id)
    player_service.assert_owner(player, user)
    session = await player_service.add_test_session(db, player_id, body)
    return TestSessionResponse(
        id=session.id,
        player_id=session.player_id,
        recorded_at=session.recorded_at.isoformat(),
        category=session.category,
        test_name=session.test_name,
        sprint_20m_fwd=session.sprint_20m_fwd,
        sprint_20m_bwd=session.sprint_20m_bwd,
        sprint_60m=session.sprint_60m,
        standing_jump=session.standing_jump,
        long_jump=session.long_jump,
        agility=session.agility,
        flexibility=session.flexibility,
        push_ups=session.push_ups,
        pull_ups=session.pull_ups,
        plank_sec=session.plank_sec,
        balance_test_sec=session.balance_test_sec,
    )


@router.get("/{player_id}/tests", response_model=list[TestSessionResponse])
async def list_test_sessions(
    player_id: str, db: AsyncSession = Depends(get_db)
):
    """List all physical test sessions for a player."""
    sessions = await player_repo.list_test_sessions(db, player_id)
    return [
        TestSessionResponse(
            id=s.id,
            player_id=s.player_id,
            recorded_at=s.recorded_at.isoformat(),
            category=s.category,
            test_name=s.test_name,
            sprint_20m_fwd=s.sprint_20m_fwd,
            sprint_20m_bwd=s.sprint_20m_bwd,
            sprint_60m=s.sprint_60m,
            standing_jump=s.standing_jump,
            long_jump=s.long_jump,
            agility=s.agility,
            flexibility=s.flexibility,
            push_ups=s.push_ups,
            pull_ups=s.pull_ups,
            plank_sec=s.plank_sec,
            balance_test_sec=s.balance_test_sec,
        )
        for s in sessions
    ]
